### DISCUSSION TITLE
Why is 1.24 ESP32-s3 PWM wrong?   October 26 git

### DISCUSSION THREAD

**Discussion by Ethanlaucan** [micropython/micropython#16086] [ESP32] [answered]
version to 1.24.0
When I set the PWM pin, Thooy says ValueError: unreachable frequency 5000

When I roll back version to 1.23.0
Everything is fine

I'm using the ESP32_GENERIC_S3 build, esp-idf-V5.2.2

I tried to build using the file downloaded on September 30, PWM is no problem  -- MicroPython v1.24.0-preview.363.g17d823445

**Comment by andrewleech** [comment]
When you try it on the previous version, do you have any way to measure the actual frequency generated on the pin? Eg an oscilloscope? 

I ask because there was a bug found in the previous versions where the clock sources used to run the pwm were declared incorrectly, I expect the frequency was not actually calculated correctly in the older version.

I'm not sure if that's what's going on here but it might be related.

**Reply by Ethanlaucan** [reply]
When I run the code in MicroPython v1.24.0-preview.363.g17d823445
This code works normally and the actual frequency can be read accurately on the oscilloscope.
{
from machine import PWM,Pin
pwm= PWM(6)
pwm.duty_u16(int(50*655.35))
pwm.freq(200)#1000/2000/25000 can be read accurately by an oscilloscope.
}

When I run the version MicroPython v1.24.0-preview.447.g838f21298
Almost the same code used
{
from machine import PWM,Pin
import time
#help(PWM)
pwm= PWM(6)
pwm.duty_u16(int(50*655.35))
pwm.freq(200)
}
But the tip Thooy gave me was this:
MPY: soft reboot
Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
ValueError: unreachable frequency 5000

Unable to read frequency on oscilloscope

**Reply by Ethanlaucan** [reply]
I tried pulling the 1.24 version updated on October 25th, but it was the same error as MicroPython v1.24.0-preview.447.g838f21298

**Comment by robert-hh** [comment, marked as answer]
It is this commit which breaks it:
6d799378b `2024-10-09 esp32: Add support for esp32c6.`
And especially these lines, causing the frequency to be limited to a range of 4 - 2000 Hz.
```
        #if SOC_LEDC_SUPPORT_XTAL_CLOCK
        timer->clk_cfg = LEDC_USE_XTAL_CLK;
        #else
        timer->clk_cfg = LEDC_USE_APB_CLK;
        #endif
```
`SOC_LEDC_SUPPORT_XTAL_CLOCK` is defined for all chips except ESP32.
That difference is strange since the S3 board I have uses a 40MHz Crystal, which is of course different to the 80Mhz APB clock. But still 5 kHz should be possible.

**Reply by Ethanlaucan** [reply]
Will there be a branch or alternative build for this problem in the future?

**Reply by projectgus** [reply]
@Ethanlaucan It's in a separate reply below, but Andrew has submitted a fix here: https://github.com/micropython/micropython/pull/16090

Once this is merged it will go into next nightly MicroPython build.

**Comment by robert-hh** [comment]
This is a BUG and has to be fixed. The actual firmware works for ESP32 and indeed ESP32C6, but fails on ESP32S3. I did not yet test the other ESP32 variants.
@andrewleech 
edit: b.t.w. Andrew. If you're at it, would you enable REPL over the built-in USB CDC? At the moment it's UART only.

**Comment by andrewleech** [comment]
> would you enable REPL over the built-in USB CDC

Is this on the S3? Which board profile? Repl was enabled on USB there last I checked, my main S3 board doesn't have UART wired at all. 

There has been a couple of changes more recently enough to make it easier to build without USB enabled to free up the pins, maybe the defaults for accidentally changed then.

**Reply by Ethanlaucan** [reply]
I am using the ESP32_GENERIC_S3 build and only modify the #define MICROPY_HW_ENABLE_UART_REPL (0) of ESP32_GENERIC_S3/mpconfigboard.h

Keep other settings as default and build the bin file configured for this board.

**Reply by Ethanlaucan** [reply]
I also saw the instructions for releasing the pins occupied by USB, but since I didn't find any relevant cases on how to modify it, I haven't modified other configurations yet.
Disabling USB is also a function I need. I hope to be able to protect the built-in code to a certain extent by disabling USB. There is currently no good OTA upgrade code solution.

**Reply by robert-hh** [reply]
The comment about REPL with USB CDC was for the ESP32C6 version. USB Repl works fine with a S3.

**Reply by andrewleech** [reply]
> The comment about REPL with USB CDC was for the ESP32C6 version. USB Repl works fine with a S3.
> 

Oh strange, similarly my only C6 module only has the USB port wired up, not UART. So USB repl was working fine for me. I'll have to test it again.

**Reply by robert-hh** [reply]
I wired up the UART with an external CH340 module. The USB port shows a CDC device, but there is no REPL prompt at it. lsusb shows:
`ID 303a:1001 Espressif USB JTAG/serial debug unit`

**Reply by andrewleech** [reply]
>  The USB port shows a CDC device, but there is no REPL prompt at it. lsusb shows:
> `ID 303a:1001 Espressif USB JTAG/serial debug unit`

I wasn't really able to reproduce this today. 

I compiled v1.24.0 and a few commits either side. 
The first time I flashed a build from v1.24.0 I connected to USB com and had no repl but I think it might have just been stuck in bootloader mode? 
I dropped back to some older commits, build and deploy, they all worked. Then went to master, and back to v1.24.0, build and deploy, all worked fine then. Each time I confirmed the commit running with a ctrl-d 

The USB port does look exactly the same to computer in both boot and normal mode on this chip, the USB connection doesn't even reset when the micro is reset.
It pretty much acts just like an external USB serial chip really, just without the UART in-between.

**Reply by robert-hh** [reply]
> the USB connection doesn't even reset when the micro is reset.

That's different here. When I reset the board, the connection is dropped. A few more observations:

1. I used to connect the board through a USB hub. When I connect the board directly to the PC, I have REPL at USB. At the hub I had it once today and I recall that it worked once a few days ago. I have another tiny hub where it works fine.
2. Every time I connect the board, the device number shown in lsusb increases. At the PC it increase by 1, at the hub it increases by 3.
3. Both at the PC and the tiny hub I see the output echoing at startup which happened before. It seems as if the ">>>" prompt is echoed.
```
Traceback (most recent call last):
  File "<stdin>", line 1
SyntaxError: invalid syntax
>>> 
>>> >>> 
Traceback (most recent call last):
  File "<stdin>", line 1
SyntaxError: invalid syntax
>>> 
>>> >>> 
Traceback (most recent call last):
  File "<stdin>", line 1
SyntaxError: invalid syntax
>>> 
>>> >>>> 
Traceback (most recent call last):
  File "<stdin>", line 1
SyntaxError: invalid syntax
>>>
```
The board I have is from WEACT ESP32C6-N4 Mini V1.0.  And it carries definitely only the bare C6 chip, regulators, switches and LEDs. No additional components, no external flash, no USB/UART bridge.

**Comment by andrewleech** [comment]
> ValueError: unreachable frequency

Ok, so that error message is given for _any_ error returned by https://github.com/micropython/micropython/blob/6d799378bad4474e77ddb2fa2187ecd6e290e0ba/ports/esp32/machine_pwm.c#L256

I'm not sure why the new settings are wrong here for S3, though there are multiple issues in this pwm configuration code, eg. https://github.com/micropython/micropython/pull/15862#discussion_r1809990023

I'm in the middle of USB work for other ports, not sure when I'll have time to look at esp32 next. 

Perhaps there's some newer reference code elsewhere for setting up pwm on all the esp32 chips we could use as a guide for doing this better.

**Comment by andrewleech** [comment]
I've got a fix for the issue raised in https://github.com/micropython/micropython/pull/16090 based on previous suggestions by @projectgus
Only tested on S3 myself, I'd be very grateful if anyone can test on other parts.

@robert-hh This test was using repl over usb cdc (using the full / tinyusb based usb stack) which is default on ESP32_GENERIC_S3. Or were you referring to the alternate USJ cdc stack in the chip which can alternatively be enabled instead? I haven't tried using that on the S3.

**Reply by robert-hh** [reply]
Will do today with esp32 generic, c3, s2, s3, c6

**Reply by robert-hh** [reply]
Tested with ESP32 Geenric, ESP32S2, ESP32S3, ESP32C3 and ESP32C6. Setting frequency to 10Hz, 1250Hz, 5kHz, 10MHz. All work fine with this Fix.

**Reply by andrewleech** [reply]
Thanks @robert-hh that's fantastic!

**Reply by robert-hh** [reply]
Forgot to say: all builds were made with esp-idf 5.2.2.

**Reply by dpgeorge** [reply]
Very good, thanks @robert-hh for the thorough testing.
